[DM-36073] First draft of spawner design

rra opened the pull request:
The first draft of a design for replacing Kubespawner with a separate spawner service that JupyterHub talks to via a REST API. This service would incorporate the functionality of moneypenny and cachemachine and take advantage of Rubin-specific knowledge to simplify lab configuration and spawning.

rra committed:
First draft of remote spawner specification

Document a basic spawner REST API and the necessary glue to make
it work with the JupyterHub Spawner API.  Describe the pod
configuration environment based on a modified version of current
practice.

Update the build system to the latest template.

rra committed:
Add prepulling and future work

Add a description of prepulling, and add a future work section.
Clean up the permission model to make it possible to implement
with current Gafaelfawr and remove some unnecessary complexity.

rra committed:
Fix syntax errors

rra committed:
Add more concurrency details for progress

Be more specific about how progress is implemented.

rra committed:
Rename APISpawner to RESTSpawner

athornton commented on index.rst:
"that broad of permissions" -> "such a broad set of permissions"

rra committed:
Update architecture diagram

Use the real ConfigMap and Secret resources that will be created by
the spawner, and rename APISpawner to RESTSpawner.

rra committed:
Drop section numbering

This doesn't seem all that helpful for this technote.

jonathansick reviewed (commented):
From my initial reading this makes a lot of sense, and will provide a lot of benefits for us. The only sketchy part is the integration within JupyterHub itself, but hopefully that'll go smoothly.

Some initial comments:

- Section 2 doesn't mention the culler as an ancillary service; but perhaps it should be considered a service in this ecosystem, and perhaps a feature of this new service?
- I've never liked Jupyter's colourful "spawner" terminology. In Kubernetes we'd usually call this an "operator." I get, though, if we need to keep using "spawner" for consistency with the Jupyter ecosystem. However, naming the service "Spawner" and giving it a `/spawner/` API path might be a little generic? I almost wonder if the service we're talking about *is the new nublado?* I.e. `/nublado/`?
- For `GET /spawner/v1/labs/<username>` it's be great if there was a version of this endpoint available to the user with `exec:notebook` rather than just `admin:notebook`. For example, it'd be great if Squareone could indicate to a user if they're currently running a pod, and what image version and size it's using. This could provide a nicer integrated experience.

jonathansick commented:
A reminder from the co-work last week that we should keep in mind the headless Nublado use case of Noteburst (Times Square). If we can simplify in in any way that'd be great (a reminder that this is what Times Square currently does, based on Mobu: https://github.com/lsst-sqre/noteburst/blob/main/src/noteburst/jupyterclient/jupyterlab.py)

Note that Times Square does call cachemachine to determine the resolve the image names at https://github.com/lsst-sqre/noteburst/blob/e09a6413227032052f2c06defd6199f010e455a7/src/noteburst/jupyterclient/jupyterlab.py#L538-L553
There are use cases for resolving the recommended image, the latest weekly and daily images, and also a specific image (i.e. if noteburst is running a worker for a boutique image for testing or some other use case)

rra committed:
Use POST not PUT for lab creation in diagram

rra committed:
Add a user-status route

Add a route that allows a user to retrieve the status of their own
pod if it is running.  This may be used by squareone to show more
details in the top-level UI.

rra committed:
Add information about scaling

Scaling is going to be rather complicated, so we don't want to
tackle it at first, but we will want it eventually.  Sketch out a
design for how it can be achieved.

rra committed:
Document a lab NetworkPolicy

We should create a NetworkPolicy for the lab with ingress and
egress controls for defense in depth.  Add a section about that
and add it to the diagram.

rra commented:
My understanding is that the culler is built-in functionality in JupyterHub and we don't need to implement it or interact with it in any way. (When it decides the pod has run for too long, JupterHub will just shut down the pod in the normal way, as if the user had stopped it.) @athornton, is that correct?

I'm not tied to the spawner naming, and will go ahead and change that. I think technically this is a Kubernetes controller but not an operator, since it doesn't use custom resources or a processing loop. But that does mean we need a new prefix for the routes. Using `/notebook` when the UI for the user uses `/nb` seems like it might be a bit confusing, and likewise for `/nublado` since nublado2 is the name of the JupyterHub service. And because `/labs` is an obvious resource tree, that rules out something like that. I'm not sure what name would naturally include notebook creation and image pre-pulling that isn't `/notebook` or `/nublado`.

After thinking about this for a bit, I think the best way of handling the noteburst case is to define `POST` parameters to the JupyterHub spawn page and have it continue to work through JupyterHub so that it manages the labs and their authentication. That saves having to reimplement authentication for noteburst. Since JupyterHub will just pass any POST parameters straight to the spawner, we can say that the following parameters will always exist and be sufficient to spawn a lab regardless of what we do with the spawner form, and define that as an API.

rra committed:
Fix terminology to avoid "spawn"

JupyterHub uses "spawn" but the standard terminology in Kubernetes
is a controller and creating resources or starting containers.  Be
consistent with that terminology, using "spawn" only for the
JupyterHub class and its API.

rra committed:
Add API to show known images

Make it easier to see what the latest recommended, weekly, daily,
and release are, and get a complete list of images that isn't
divided into what's being prepulled.  Remove the list of images
that aren't being prepulled from the prepull status.

rra committed:
Simplify the prepull configuration

Rather than keeping the idea of multiple prepull configurations,
elevate the prepull configuration to the top level and assume
Rubin tag conventions.  We're not trying to be a generic prepuller,
we're specifically a controller for labs for the Notebook Aspect of
the Rubin Science Platform, so can take a much simpler approach.

rra committed:
Add proposal for lab creation for bot users

Add a mechanism for bot users to create labs via a POST to
JupyterHub without caring about what fields are present in the
lab creation form.

Also change the prepuller configuration to be a generic
configuration for what images are available to the Notebook
Aspect, which also happens to control prepulling.

athornton commented on index.rst:
"emitting"

athornton commented on index.rst:
May want to specify what happens if both are set.  I think image_tag should win.

athornton commented on index.rst:
In the most common case, one of "small", "medium", or"large".  It has been our practice thus far that each size is double the CPU and memory of the size immediately below it, and we have typically started with 3GB or 4GB per CPU.  The sizes come from the d20 SRD for creature size: https://dungeons.fandom.com/wiki/SRD:Table_of_Creature_Size_and_Scale .

athornton commented on index.rst:
I realize why this is literal text but it would look better as "Network Policy".

How do we define "external IP addresses"?  If you just mean "not-RFC-1918 space" then we also (I think) need an exception for our storage systems (at least at the IDF I think these are addressed through non-public IP ranges).

athornton reviewed (approved):
I adore this.  I have a few nits, but nothing substantive.  This seems like the right way forward.

rra committed:
Fixes from review

- Fix typos and spelling errors
- Clarify non-cluster IP addresses in network policy section
- Specify that image_tag overrides image_type

rra committed:
Change URL prefix to /nublado/spawner

We'd prefer /nb/spawner, but then it tromps on the JupyterHub
route namespace, which may be a problem.
